algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py

```diff
@@ -1055,44 +1401,52 @@ def multi_head_attention_forward(
     assert bias_k is None
     assert bias_v is None
 
-  #
-  # reshape q, k, v for multihead attention and make em batch first
-  #
-  q = q.contiguous().view(tgt_len, bsz * num_heads, head_dim).transpose(0, 1)
-  k = k.contiguous().view(k.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
-  v = v.contiguous().view(v.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
+  # Reshape q, k, v for multihead attention and make em batch first.
+  q = \
+    q.contiguous().view(tgt_len, bsz * num_heads, head_dim).transpose(0, 1)
+  k = \
+    k.contiguous().view(k.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
+  v = \
+    v.contiguous().view(v.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
 
-  # update source sequence length after adjustments
+  # Update source sequence length after adjustments.
   src_len = k.size(1)
 
-  # convert mask to float
+  # Convert mask to float.
   if attn_mask is not None and attn_mask.dtype == torch.bool:
     new_attn_mask = torch.zeros_like(attn_mask, dtype=q.dtype)
     new_attn_mask.masked_fill_(attn_mask, -1e10)
     attn_mask = new_attn_mask
 
-  # adjust dropout_rate probability
+  # Adjust dropout_rate probability.
   if not training:
     dropout_rate = 0.0
 
-  #
-  # (deep breath) calculate attention and out projection
-  #
-  attn_output, attn_output_weights = _scaled_dot_product_attention(
+  # Calculate attention and out projection.
+  attn_output = torch.nn.functional.scaled_dot_product_attention(
       q, k, v, attn_mask, dropout_rate)
   attn_output = attn_output.transpose(0, 1).contiguous().view(
       tgt_len * bsz, embed_dim)
   attn_output = F.linear(attn_output, out_proj_weight, out_proj_bias)
   attn_output = attn_output.view(tgt_len, bsz, attn_output.size(1))
 
   if need_weights:
-    # optionally average attention weights over heads
+    q_scaled = q / math.sqrt(q.shape[-1])
+
+    if attn_mask is not None:
+      attn_output_weights = torch.baddbmm(attn_mask,
+                                          q_scaled,
+                                          k.transpose(-2, -1))
+    else:
+      attn_output_weights = torch.bmm(q_scaled, k.transpose(-2, -1))
+
+    # Optionally average attention weights over heads.
     attn_output_weights = attn_output_weights.view(bsz,
                                                    num_heads,
                                                    tgt_len,
                                                    src_len)
     if average_attn_weights:
-      attn_output_weights = attn_output_weights.sum(dim=1) / num_heads
+      attn_output_weights = attn_output_weights.mean(dim=1) / num_heads
```

Comment:
What was the motivation for the change here? If `dim=1` is the dimension over the heads and we switch to `mean` instead of `sum`, the division by `num_heads` should be removed or not?

Reply:
My bad! I will fix this.